src/hooks/useSelector.js

```diff
@@ -20,13 +20,24 @@ function useSelectorWithStoreAndSubscription(
   ])
 
   const latestSubscriptionCallbackError = useRef()
-  const latestSelector = useRef()
-  const latestStoreState = useRef()
-  const latestSelectedState = useRef()
 
   const storeState = store.getState()
+
+  const latestSelector = useRef(selector)
+  const latestStoreState = useRef(storeState)
+  const latestSelectedState = useRef()
+
   let selectedState
 
+  // lazy initialization pattern on first render
+  if (latestSelectedState.current === undefined) {
+    selectedState = selector(storeState)
+    latestSelectedState.current = {
+      first: true, // mark that this is the first render, so we can avoid calling selector in checkForUpdates
+      val: selectedState,
+    }
+  }
+
```

Comment:
By initializing refs here (lazily for the `latestSelectedState` since it can be complex) and by value for `latestSelector` / `latestStoreState`, we can separate the generalized (if anything changes, call the selector again) recalculation below: https://github.com/sufian-slack/react-redux/pull/1/files#diff-19c189d7f6bf5e0a2646567980762c0b615e6cdeee0079ca781bb10c5cb92198R47